[ICE]: Enum discriminant referring to `Self` causes "... has parameters, but no args were provided in instantiate"

### Code

```Rust
#![feature(sized_hierarchy)]

use std::marker::PointeeSized;

#[repr(usize)]
enum What<T: PointeeSized> {
    X = size_of::<*mut Self>(),
    Y(*mut T)
}
```

### Meta

Reproducible on the playground with version `1.96.0-nightly (2026-03-11 3b1b0ef4d80d3117924d)`

### Error output

```
thread 'rustc' (27) panicked at /rustc-dev/3b1b0ef4d80d3117924d91352c8b6ca528708b3c/compiler/rustc_type_ir/src/binder.rs:658:13:
Val(ZeroSized, FnDef(DefId(2:2522 ~ core[d6e6]::mem::size_of), [*mut What<T/#0>])) has parameters, but no args were provided in instantiate
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: <rustc_const_eval::interpret::eval_context::InterpCx<rustc_const_eval::const_eval::machine::CompileTimeMachine>>::eval_operand
   3: <rustc_const_eval::interpret::eval_context::InterpCx<rustc_const_eval::const_eval::machine::CompileTimeMachine>>::eval_callee_and_args
   4: rustc_const_eval::const_eval::eval_queries::eval_to_allocation_raw_provider
      [... omitted 1 frame ...]
   5: rustc_const_eval::const_eval::eval_queries::eval_to_const_value_raw_provider
      [... omitted 1 frame ...]
   6: <rustc_middle::ty::context::TyCtxt>::const_eval_poly
   7: <rustc_middle::ty::adt::AdtDef>::eval_explicit_discr
   8: rustc_hir_analysis::check::check::check_item_type
   9: rustc_hir_analysis::check::wfcheck::check_well_formed
      [... omitted 1 frame ...]
  10: rustc_hir_analysis::check::wfcheck::check_type_wf
      [... omitted 1 frame ...]
  11: rustc_hir_analysis::check_crate
  12: rustc_interface::passes::analysis
  13: rustc_query_impl::execution::try_execute_query::<rustc_middle::query::caches::SingleCache<rustc_middle::query::erase::ErasedData<[u8; 0]>>, false>
  14: rustc_interface::interface::run_compiler::<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: the compiler unexpectedly panicked. This is a bug

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/playground/rustc-ice-2026-03-12T04_25_48-25.txt` to your bug report

note: rustc 1.96.0-nightly (3b1b0ef4d 2026-03-11) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type lib -C embed-bitcode=no -C codegen-units=1 -C debuginfo=2

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [eval_to_allocation_raw] const-evaluating + checking `What::X::{constant#0}`
#1 [eval_to_const_value_raw] simplifying constant for the type system `What::X::{constant#0}`
#2 [check_well_formed] checking that `What` is well-formed
#3 [check_type_wf] checking that types are well-formed
#4 [analysis] running analysis passes on crate `playground`
end of query stack
```

</p>
</details>

Some other odd cases:

This errors
```rust
#[repr(usize)]
enum What<T: ?Sized> {
    X = size_of::<*mut Self>(),
    Y(*mut T)
}
```
```
rror[E0277]: the size for values of type `T` cannot be known
 --> src/lib.rs:3:19
  |
3 |     X = size_of::<*mut Self>(),
  |                   ^^^^^^^^^ doesn't have a known size
  |
note: required by a bound in `What`
 --> src/lib.rs:2:11
  |
2 | enum What<T: ?Sized> {
  |           ^ required by this bound in `What`
help: consider further restricting type parameter `T` with unstable trait `MetaSized`
  |
2 | enum What<T: ?Sized + std::marker::MetaSized> {
  |                     ++++++++++++++++++++++++

For more information about this error, try `rustc --explain E0277`.
```

This compiles
```rust
#[repr(usize)]
enum What {
    X = size_of::<*mut Self>(),
}
```

This errors
```rust
#[repr(usize)]
enum What {
    X = size_of::<Self>(),
}
```
```
error[E0391]: cycle detected when simplifying constant for the type system `What::X::{constant#0}`
    --> src/lib.rs:3:9
     |
   3 |     X = size_of::<Self>(),
     |         ^^^^^^^^^^^^^^^^^
     |
note: ...which requires const-evaluating + checking `What::X::{constant#0}`...
    --> /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/mem/mod.rs:345:5
     |
 345 |     <T as SizedTypeProperties>::SIZE
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires simplifying constant for the type system `core::mem::SizedTypeProperties::SIZE`...
    --> /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/mem/mod.rs:1254:5
     |
1254 |     const SIZE: usize = intrinsics::size_of::<Self>();
     |     ^^^^^^^^^^^^^^^^^
note: ...which requires const-evaluating + checking `core::mem::SizedTypeProperties::SIZE`...
    --> /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/mem/mod.rs:1254:5
     |
1254 |     const SIZE: usize = intrinsics::size_of::<Self>();
     |     ^^^^^^^^^^^^^^^^^
     = note: ...which requires computing layout of `What`...
     = note: ...which again requires simplifying constant for the type system `What::X::{constant#0}`, completing the cycle
note: cycle used when checking that `What` is well-formed
    --> src/lib.rs:2:1
     |
   2 | enum What {
     | ^^^^^^^^^
     = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information

For more information about this error, try `rustc --explain E0391`.
```

--- a/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs
+++ b/compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs
@@ -376,24 +376,44 @@ pub trait GenericArgsLowerer<'a, 'tcx> {
     ) -> ty::GenericArg<'tcx>;
 }
 
-struct ForbidMCGParamUsesFolder<'tcx> {
+/// Context in which `ForbidParamUsesFolder` is being used, to emit appropriate diagnostics.
+enum ForbidParamContext {
+    /// Anon const in a const argument position.
+    ConstArgument,
+    /// Enum discriminant expression.
+    EnumDiscriminant,
+}
+
+struct ForbidParamUsesFolder<'tcx> {
     tcx: TyCtxt<'tcx>,
     anon_const_def_id: LocalDefId,
     span: Span,
     is_self_alias: bool,
+    context: ForbidParamContext,
 }
 
-impl<'tcx> ForbidMCGParamUsesFolder<'tcx> {
+impl<'tcx> ForbidParamUsesFolder<'tcx> {
     fn error(&self) -> ErrorGuaranteed {
-        let msg = if self.is_self_alias {
-            "generic `Self` types are currently not permitted in anonymous constants"
-        } else if self.tcx.features().generic_const_args() {
-            "generic parameters in const blocks are only allowed as the direct value of a `type const`"
-        } else {
-            "generic parameters may not be used in const operations"
+        let msg = match self.context {
+            ForbidParamContext::EnumDiscriminant if self.is_self_alias => {
+                "generic `Self` types are not permitted in enum discriminant values"
+            }
+            ForbidParamContext::EnumDiscriminant => {
+                "generic parameters may not be used in enum discriminant values"
+            }
+            ForbidParamContext::ConstArgument if self.is_self_alias => {
+                "generic `Self` types are currently not permitted in anonymous constants"
+            }
+            ForbidParamContext::ConstArgument => {
+                if self.tcx.features().generic_const_args() {
+                    "generic parameters in const blocks are only allowed as the direct value of a `type const`"
+                } else {
+                    "generic parameters may not be used in const operations"
+                }
+            }
         };
         let mut diag = self.tcx.dcx().struct_span_err(self.span, msg);
-        if self.is_self_alias {
+        if self.is_self_alias && matches!(self.context, ForbidParamContext::ConstArgument) {
             let anon_const_hir_id: HirId = HirId::make_owner(self.anon_const_def_id);
             let parent_impl = self.tcx.hir_parent_owner_iter(anon_const_hir_id).find_map(
                 |(_, node)| match node {
@@ -407,18 +427,20 @@ impl<'tcx> ForbidMCGParamUsesFolder<'tcx> {
                 diag.span_note(impl_.self_ty.span, "not a concrete type");
             }
         }
-        if self.tcx.features().min_generic_const_args() {
+        if matches!(self.context, ForbidParamContext::ConstArgument)
+            && self.tcx.features().min_generic_const_args()
+        {
             if !self.tcx.features().generic_const_args() {
                 diag.help("add `#![feature(generic_const_args)]` to allow generic expressions as the RHS of const items");
             } else {
                 diag.help("consider factoring the expression into a `type const` item and use it as the const argument instead");
             }
-        };
+        }
         diag.emit()
     }
 }
 
-impl<'tcx> ty::TypeFolder<TyCtxt<'tcx>> for ForbidMCGParamUsesFolder<'tcx> {
+impl<'tcx> ty::TypeFolder<TyCtxt<'tcx>> for ForbidParamUsesFolder<'tcx> {
     fn cx(&self) -> TyCtxt<'tcx> {
         self.tcx
     }
@@ -460,37 +482,80 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
             && tcx.def_kind(parent_def_id) == DefKind::AnonConst
             && let ty::AnonConstKind::MCG = tcx.anon_const_kind(parent_def_id)
         {
-            let folder = ForbidMCGParamUsesFolder {
+            let folder = ForbidParamUsesFolder {
                 tcx,
                 anon_const_def_id: parent_def_id,
                 span,
                 is_self_alias: false,
+                context: ForbidParamContext::ConstArgument,
             };
             return Err(folder.error());
         }
         Ok(())
     }
 
+    /// Returns the `ForbidParamContext` for the current anon const if it is a context that
+    /// forbids uses of generic parameters. `None` if the current item is not such a context.
+    ///
+    /// Name resolution handles most invalid generic parameter uses in these contexts, but it
+    /// cannot reject `Self` that aliases a generic type, nor generic parameters introduced by
+    /// type-dependent name resolution (e.g. `<Self as Trait>::Assoc` resolving to a type that
+    /// contains params). Those cases are handled by `check_param_uses_if_mcg`.
+    fn anon_const_forbids_generic_params(&self) -> Option<ForbidParamContext> {
+        let tcx = self.tcx();
+        let parent_def_id = self.item_def_id();
+
+        // Inline consts and closures can be nested inside anon consts that forbid generic
+        // params (e.g. an enum discriminant). Walk up the def parent chain to find the
+        // nearest enclosing AnonConst and use that to determine the context.
+        let anon_const_def_id = match tcx.def_kind(parent_def_id) {
+            DefKind::AnonConst => parent_def_id,
+            DefKind::InlineConst | DefKind::Closure => {
+                let root = tcx.typeck_root_def_id(parent_def_id.into());
+                match tcx.def_kind(root) {
+                    DefKind::AnonConst => root.expect_local(),
+                    _ => return None,
+                }
+            }
+            _ => return None,
+        };
+
+        match tcx.anon_const_kind(anon_const_def_id) {
+            ty::AnonConstKind::MCG => Some(ForbidParamContext::ConstArgument),
+            ty::AnonConstKind::NonTypeSystem => {
+                // NonTypeSystem anon consts only have accessible generic parameters in specific
+                // positions (ty patterns and field defaults — see `generics_of`). In all other
+                // positions (e.g. enum discriminants) generic parameters are not in scope.
+                if tcx.generics_of(anon_const_def_id).count() == 0 {
+                    Some(ForbidParamContext::EnumDiscriminant)
+                } else {
+                    None
+                }
+            }
+            ty::AnonConstKind::GCE
+            | ty::AnonConstKind::GCA
+            | ty::AnonConstKind::RepeatExprCount => None,
+        }
+    }
+
     /// Check for uses of generic parameters that are not in scope due to this being
-    /// in a non-generic anon const context.
+    /// in a non-generic anon const context (e.g. MCG or an enum discriminant).
+    ///
+    /// Name resolution rejects most invalid uses, but cannot handle `Self` aliasing a
+    /// generic type or generic parameters introduced by type-dependent name resolution.
     #[must_use = "need to use transformed output"]
     fn check_param_uses_if_mcg<T>(&self, term: T, span: Span, is_self_alias: bool) -> T
     where
         T: ty::TypeFoldable<TyCtxt<'tcx>>,
     {
         let tcx = self.tcx();
-        let parent_def_id = self.item_def_id();
-        if tcx.def_kind(parent_def_id) == DefKind::AnonConst
-            && let ty::AnonConstKind::MCG = tcx.anon_const_kind(parent_def_id)
+        if let Some(context) = self.anon_const_forbids_generic_params()
             // Fast path if contains no params/escaping bound vars.
             && (term.has_param() || term.has_escaping_bound_vars())
         {
-            let mut folder = ForbidMCGParamUsesFolder {
-                tcx,
-                anon_const_def_id: parent_def_id,
-                span,
-                is_self_alias,
-            };
+            let anon_const_def_id = self.item_def_id();
+            let mut folder =
+                ForbidParamUsesFolder { tcx, anon_const_def_id, span, is_self_alias, context };
             term.fold_with(&mut folder)
         } else {
             term

--- a/tests/ui/enum-discriminant/generic-self-in-discr-ice.rs
+++ b/tests/ui/enum-discriminant/generic-self-in-discr-ice.rs
@@ -1,0 +1,12 @@
+#![feature(sized_hierarchy)]
+
+use std::marker::PointeeSized;
+
+#[repr(usize)]
+enum What<T: PointeeSized> {
+    X = size_of::<*mut Self>(),
+    //~^ ERROR generic `Self` types are not permitted in enum discriminant values
+    Y(*mut T),
+}
+
+fn main() {}

--- a/tests/ui/enum-discriminant/generic-self-in-discr-ice.stderr
+++ b/tests/ui/enum-discriminant/generic-self-in-discr-ice.stderr
@@ -1,0 +1,8 @@
+error: generic `Self` types are not permitted in enum discriminant values
+  --> $DIR/generic-self-in-discr-ice.rs:7:24
+   |
+LL |     X = size_of::<*mut Self>(),
+   |                        ^^^^
+
+error: aborting due to 1 previous error
+

--- a/tests/ui/enum-discriminant/generic-self-in-discr-inline-const.rs
+++ b/tests/ui/enum-discriminant/generic-self-in-discr-inline-const.rs
@@ -1,0 +1,17 @@
+//@ check-fail
+// Test that `Self` is rejected even when nested inside an inline const
+// or closure within an enum discriminant. Regression test for issue #154281.
+#![feature(sized_hierarchy)]
+
+use std::marker::PointeeSized;
+
+#[repr(usize)]
+enum What<T: PointeeSized> {
+    X = const { { let _: *mut Self; 1_usize } },
+    //~^ ERROR generic `Self` types are not permitted in enum discriminant values
+    Y = { let _f = || { let _: *mut Self; }; 1_usize },
+    //~^ ERROR generic `Self` types are not permitted in enum discriminant values
+    Z(*mut T),
+}
+
+fn main() {}

--- a/tests/ui/enum-discriminant/generic-self-in-discr-inline-const.stderr
+++ b/tests/ui/enum-discriminant/generic-self-in-discr-inline-const.stderr
@@ -1,0 +1,14 @@
+error: generic `Self` types are not permitted in enum discriminant values
+  --> $DIR/generic-self-in-discr-inline-const.rs:10:31
+   |
+LL |     X = const { { let _: *mut Self; 1_usize } },
+   |                               ^^^^
+
+error: generic `Self` types are not permitted in enum discriminant values
+  --> $DIR/generic-self-in-discr-inline-const.rs:12:37
+   |
+LL |     Y = { let _f = || { let _: *mut Self; }; 1_usize },
+   |                                     ^^^^
+
+error: aborting due to 2 previous errors
+

